### DISCUSSION TITLE
Making a GET-request with Wiznet Ethernet Hat

### DISCUSSION THREAD

**Discussion by LavaCode** [micropython/micropython#14112] [RP2040 / Pico]
Hi all, 

I have a Raspberry Pi Pico with the Wiznet ethernet hat and trying to make very basic HTTP-request to a connected computer. 
The computer is running Madrix Software ("regular" software application) which has a webserver to be controlled remotely. 

IP Config:
Server (Madrix): 10.0.0.3 port 80
Client (Pico/Wiznet): 10.0.0.2 

The server and client are directly connected - 15cm LAN cable. 

All this works fine, but I can't get the Wiznet board working.. 
Due to a direct connection to the server I want to have the board to have a static IP. 
Which is more reliable. 

So loading this script (which executes a single request, for now) will give me an error. 

```
from usocket import socket
from machine import Pin,SPI
import urequests
import network
import time

#W5x00 chip init
def w5x00_init():
    spi=SPI(0,2_000_000, mosi=Pin(19),miso=Pin(16),sck=Pin(18))
    nic = network.WIZNET5K(spi,Pin(17),Pin(20)) #spi,cs,reset pin
    nic.ifconfig(('10.0.0.2','255.0.0.0','10.0.0.1','8.8.8.8'))
    nic.active(True)
    
#None DHCP
    nic.ifconfig(('10.0.0.2','255.0.0.0','10.0.0.1','8.8.8.8'))
    
#DHCP
    #nic.ifconfig('dhcp')
#     print('IP address :', nic.ifconfig())
#     
#     while not nic.isconnected():
#         time.sleep(1)
#         print(nic.regs())
    
def request():
    r = urequests.get('http://10.0.0.3/RemoteCommands/SetFadeTime=8')
    #r.raise_for_status
    print(r.status_code)
    print(r.text)


def main():
    w5x00_init()
    request()

if __name__ == "__main__":
    main()
```

I get this error:
```
MPY: soft reboot
Traceback (most recent call last):
  File "<stdin>", line 37, in <module>
  File "<stdin>", line 33, in main
  File "<stdin>", line 12, in w5x00_init
OSError: timeout waiting for DHCP to get IP address
>>> 
```

Clearly it's waiting for DHCP, but it will never get an DHCP-IP. 

Also, if I connect my board to my router instead I (and allowing the script to accept DHCP) also get's an error:
```
MPY: soft reboot
netif changed  192.168.1.179
IP address : ('192.168.1.206', '255.255.255.0', '192.168.1.1', '8.8.8.8')
Traceback (most recent call last):
  File "<stdin>", line 37, in <module>
  File "<stdin>", line 34, in main
  File "<stdin>", line 26, in request
  File "/lib/requests/__init__.py", line 180, in get
  File "/lib/requests/__init__.py", line 91, in request
OSError: [Errno 13] EACCES
>>>
```

I've read this has to do with file access? 
My aim is to not use DHCP since I want to be able to run the configuration even without any router or network. Just an interface connected to a local PC for operation.  

I'm able to make a request if I use Postman or by loading the endpoint in my browser. 
There is no authentication or anything; you just need to have connect between each other. 
What am I doing wrong here? 

Thank you

**Comment by robert-hh** [comment]
I cannot recall in which situation, but I had issues with calling nic.ifconfig() before nic.active(). So you could try to remove the one before nic.active().
About the second erro: Is in your network the address range 10.0.0.x available from 192.168.1.x?

**Reply by LavaCode** [reply]
Yes, the first nic.ifconfig() should not really be there, I've added it for testing purpose to see whether that would help, but nope. 
Unfortunately it does (also) not work if I just have the nic.ifconfig() set for a single time after nic.active() -- so with removal of the first one.

In terms of the network, no: 
the 10.0.0.x is the ethernet adapter
the 192.168.1.x is the wifi adapter. 

Madrix does enable you to control per adapter, meaning that I can control it on:
192.168.1.206 - WiFi (DHCP - internet connection) 
10.0.0.3 - Ethernet (static IP) 
127.0.0.1 - Loopback

I absolutely want to control it with static IP since the installation won't be relying on internet/WiFi. 
I also prefer to keep it separate
And I know that it is simply possible :-)

**Reply by robert-hh** [reply]
So I wired up a W5500 module to a RPI Pico with the WIZNET_EVB_5500 firmware and you code. Two observations:

0. The standard config with a DHCP server works
1. With nic.ifconfig before nic.active() the connections fails.
2. Attaching the W5500 to a network without an DHCP server it fails in the nic.active() state, but when I put the nic.active() into a try/except clause and set the IP address after the ignored fail, I have  a connection and I can communicate.

**Reply by LavaCode** [reply]
Hi @robert-hh , thank you for your help so far. 
So that helped a bit, but I still find issues. 

Currently with the updated script I do get the board on the correct IP. 
While somehow the script runs into the except, that's not supposed to happen right? (see my current script below)
But I can reach the server now, but now my issue is with the request itself. The server logs can't handle the current request due to a missing host in http header(Log: request without host in http header). The thing is I used the Pico before for doing this, which was with WiFi (and worked unstable), so my request is the same. I did some experimenting in adding headers.. 

Do you see what I'm doing wrong here? 
Current script: 
```
from usocket import socket
from machine import Pin,SPI
import urequests
import network
import time

#W5x00 chip init
def w5x00_init():
    spi=SPI(0,2_000_000, mosi=Pin(19),miso=Pin(16),sck=Pin(18))
    nic = network.WIZNET5K(spi,Pin(17),Pin(20)) #spi,cs,reset pin
    try:
        nic.active(True)
    except:
        print('error')
    
#None DHCP
    nic.ifconfig(('10.0.0.2','255.0.0.0','10.0.0.1','8.8.8.8'))
    
def request():    
    r = "http://10.0.0.3/RemoteCommands/SetFadeTime=5"
    headers = {
        "Host": "10.0.0.2",
        "accept": "application/json"
    }
    result = urequests.get(r, headers=headers)
    print("Statuscode request 1: " + str(result.status_code))
    result.close()


def main():
    w5x00_init()
    request()

if __name__ == "__main__":
    main()
```

Shell output: 
```
>>> %Run -c $EDITOR_CONTENT

MPY: soft reboot
error
Traceback (most recent call last):
  File "<stdin>", line 43, in <module>
  File "<stdin>", line 40, in main
  File "<stdin>", line 33, in request
  File "/lib/requests/__init__.py", line 180, in get
  File "/lib/requests/__init__.py", line 100, in request
OSError: -7
>>> 
```

**Reply by robert-hh** [reply]
The print of `error` in w5x00_init() was expected, because that's the one being ignored. But it seems that some step is missing. It would be interesting if that is caused by the missing DHCP step. Can you set up a temporary DHCP service in the 10.x.x.x net?

Surely the Wiznet driver needs a rework to allow starting without DHCP.

**Reply by robert-hh** [reply]
I repeated the set-up test of ignoring the error during nic.active(True). Works so far. I can run iperf3 with TCP and UPD test, and a simple TCP test I made myself. So socket communication works.

Not sure what the error -7 means. It could be "Operation would block".

**Reply by LavaCode** [reply]
Hi @robert-hh 
Yes I think in terms of communication everything works. The request is now complaining, missing a host in HTTP-header..

**Reply by robert-hh** [reply]
I've seen your logs, and compared your code to the uequests lib. I wonder why it complains about a missing host in header. According to the sources, a host field is provided.  And I just see `OSError: -7`.
Why do you think the host is missing?

**Reply by LavaCode** [reply]
![Scherm­afbeelding 2024-03-19 om 10 41 05](https://github.com/micropython/micropython/assets/68539391/b8ca3aac-3d28-4bfd-aa57-151fc080ef01)

Basically I can see it remotely on the server (application-side) 
The application does receive the request but can't handle it as it indicates in the log. 

The application not strict at all, just missing some data. 
The request is correct, since it is being recognized.

**Reply by robert-hh** [reply]
That's strange, because in urequests.py `Host` is supplied and written. I do not know whether the case "Host" vs "host" matters. You could try to verify the presend by adding a print statement into urequests.py  for the `headers` argument.
